Add Geant4 packages.

jakeforster opened the pull request:
This work started from a PR submitted by Emmanuel Medernach to guix-science: https://github.com/guix-science/guix-science/pull/25

The main things I've added and changed from that PR:
- Add geant4-vis package that supports visualisation using OGL and Qt.
- Add package for the DAWN visualisation tool.
- Switch CLHEP and Geant4 sources from archive files to version control systems so they can be archived in Software Heritage.
- Use Guix-style package names and version strings.

Thanks!
Jake

EDIT: I've added some documentation below about how to use the geant4 and geant4-vis packages.

### geant4

The package geant4 is [Geant4](https://geant4.web.cern.chl) without support for visualisation.

Versions provided:
- 11.2.1
- 11.2.0
- 11.1.3
- 11.1.2
- 11.1.1

Minimal environment:

```sh
  guix shell geant4 cmake make gcc-toolchain

  # copy an example app
  INSTALL_DIR=$(guix build geant4)
  cp -rf $INSTALL_DIR/share/Geant4/examples/basic/B1 .

  # the store is read only
  chmod -R 751 B1/

  # the usual
  cd B1
  mkdir build
  cd build
  cmake ..; make
  ./exampleB1 run1.mac
```

NB. Do not source $INSTALL_DIR/bin/geant4.sh.

### geant4-vis

The package geant4-vis is Geant4 including support for visualisation with OpenGL and Qt.

Versions provided:
- 11.2.1
- 11.2.0
- 11.1.3
- 11.1.2
- 11.1.1

Minimal environment:

```sh
  guix shell geant4-vis cmake make gcc-toolchain mesa --no-grafts

  # copy an example app
  INSTALL_DIR=$(guix build geant4-vis --no-grafts)
  cp -rf $INSTALL_DIR/share/Geant4/examples/basic/B1 .

  # the store is read only
  chmod -R 751 B1/

  # the usual
  cd B1
  mkdir build
  cd build
  cmake ..; make
  ./exampleB1
```

NB.:
- Do not source $INSTALL_DIR/bin/geant4.sh
- Use the `--no-grafts` option when installing geant4-vis and mesa. Otherwise, when using CMake to build a Geant4 application, CMake may emit a warning about not being able to generate a safe runtime search path due to there being multiple variants of mesa.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                  expat
+                  clhep-2.4.6.2
+                  python-2
+                  python-3.10
```
Both Python 2 and Python 3.10?  Is there a reason why we need 3.10 instead of the default Python?  Could you please add comments for each of these?

jakeforster replied:
It seems that many of the inputs can be removed, including python. I removed them.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (build-system cmake-build-system)
+    (inputs (list coreutils
+                  gcc-toolchain
+                  xerces-c ;pour GDML
```
Please use English for the comments :)

jakeforster replied:
Done

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+        (base32 "141fhmh0w8sbp6cckccf3dswn596ds4vgqwc3gz6i53ypyxmv2fw"))))
+    (build-system cmake-build-system)
+    (inputs (list coreutils
+                  gcc-toolchain
```
coreutils and gcc-toolchain are already among the default inputs (natively).  Is there a reason for adding them here?

jakeforster replied:
Not that I know of. I removed them.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (origin
+      (method url-fetch)
+      (uri (string-append "https://cern.ch/geant4-data/datasets/G4ENSDFSTATE."
+            version ".tar.gz"))
```
The indentation looks odd.

jakeforster replied:
Indeed. But it was output from `guix style -f`. It's no longer there because we use the geant4-dataset function as you suggested below.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+      (method url-fetch)
+      (uri (string-append "https://cern.ch/geant4-data/datasets/G4NDL."
+                          version ".tar.gz"))
+      (sha256 (base32 "0283cwylajyjm4267ngfc2bd3452623r5bakywaccb8h44k3szby")))))
```
You could have a procedure to simplify all these datasets.

```scheme
(define (geant-dataset name version hash)
  (origin
    (method url-fetch)
    (uri (string-append "https://cern.ch/geant4-data/datasets/" name "." version ".tar.gz"))
    (sha256 (base32 hash)))
```

jakeforster replied:
Fantastic! Thanks. I used this.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                  perl
+                  tcsh))
+    (arguments
+     `(#:configure-flags (let ((out (assoc-ref %outputs "out")))
```
Please use a G-expression and replace `(assoc-ref %outputs "out")` with `#$output`.

jakeforster replied:
Thanks for the suggestion. I've done this.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                  (add-after 'install 'install-data
+                    (lambda* (#:key inputs outputs #:allow-other-keys)
+                      (let ((G4NDL (assoc-ref inputs "G4NDL"))
+                            (G4EMLOW (assoc-ref inputs "G4EMLOW"))
```
Instead of `assoc-ref inputs` please use `#$(this-package-input "whatever")` or `#$(this-package-native-input "foo")`.

jakeforster replied:
Thanks. I've done this.

As a side note, I had to be careful to avoid this issue: [https://issues.guix.gnu.org/66510](https://issues.guix.gnu.org/66510)
I.e., using substitute-keyword-arguments causes this-package(-native)-input to resolve to that of the referenced package and not the current package.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                        (invoke "tar" "xvf" G4ABLA "-C" datadir)
+                        (invoke "tar" "xvf" G4INCL "-C" datadir)
+                        (invoke "tar" "xvf" G4ENSDFSTATE "-C" datadir)
+                        (invoke "tar" "xvf" G4TENDL "-C" datadir)))))
```
how about `(for-each (lambda (archive) (invoke ...) (list G4NDL G4EMLOW ...)`?

jakeforster replied:
Very good. Done.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                     ("G4ABLA" ,g4abla-3.1)
+                     ("G4INCL" ,g4incl-1.0)
+                     ("G4ENSDFSTATE" ,g4ensdfstate-2.3)
+                     ("G4TENDL" ,g4tendl-1.4)))
```
I guess we can't use the default labels here, because these are `origin`s, not `package`s.  Oh well.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+nuclear and accelerator physics, as well as studies
+in medical and space science.
+
+Note this package does not support visualisation -- you
```
"visualization"  (we use en_US for better or worse).

jakeforster replied:
Done

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+Note this package does not support visualisation -- you
+can use @code{geant4-vis} for that.")
+    (license (license:non-copyleft
+              "https://geant4.web.cern.ch/download/license"))))
```
Hmm, does this condition make the software effectively non-free?

> You may not include this software in whole or in part in any patent or patent application in respect of any modification of this software developed by you.

This has also been discussed in Debian: https://lists.debian.org/debian-legal/2021/01/msg00014.html

I can't answer this conclusively.  My recommendation would thus be to add this module to guix-science-nonfree instead, with a comment explaining why.

jakeforster replied:
Happy to send this to guix-science-nonfree instead. Feel free to close this PR and I'll open one with guix-science-nonfree with the revisions.

Added this comment: 
;; The Geant4 license has an anti-patent clause that might make it nonfree.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (inherit geant4-11-1-1)
+    (name "geant4-vis")
+    (inputs (list coreutils
+                  gcc-toolchain
```
Same comment as above.  These shouldn't be needed here.

jakeforster replied:
Done

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                  expat
+                  clhep-2.4.6.2
+                  python-2
+                  python-3.10
```
Same as above.

jakeforster replied:
Done

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                  tcsh
+                  qtbase-5
+                  libxmu
+                  libxt))
```
Since we inherit package fields already, would it not be better to use `(modify-inputs ...)` instead to augment the inputs as needed?  That would show more clearly what the effective difference between these package variants really is.

jakeforster replied:
Done. Thanks.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                                (qt-path (string-append (assoc-ref
+                                                         %build-inputs
+                                                         "qtbase")
+                                                        "/lib/cmake/Qt5")))
```
Same as above.  Please use G-expressions and replace `(assoc-ref %outputs "out")` with `#$output`, and `(assoc-ref %build-inputs ...)` with `#$(this-package-input ...)`.

jakeforster replied:
Done. Thanks

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                        (invoke "tar" "xvf" G4ABLA "-C" datadir)
+                        (invoke "tar" "xvf" G4INCL "-C" datadir)
+                        (invoke "tar" "xvf" G4ENSDFSTATE "-C" datadir)
+                        (invoke "tar" "xvf" G4TENDL "-C" datadir)))))
```
This looks all very much like the package above.  Can we inherit this stuff as well?

rekado replied:
Use `substitute-keyword-arguments` to keep some keyword arguments unchanged and only override those that have changed.

jakeforster replied:
Done. Thanks.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                  python-2
+                  python-3.10
+                  perl
+                  tcsh))
```
Same as above.  If there are changes to the inputs please use `modify-inputs`.

jakeforster replied:
Done

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                     ("G4ABLA" ,g4abla-3.3)
+                     ("G4INCL" ,g4incl-1.2)
+                     ("G4ENSDFSTATE" ,g4ensdfstate-2.3)
+                     ("G4TENDL" ,g4tendl-1.4)))))
```
Same as above.

jakeforster replied:
Done

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (version "11.2.0")
+    (source
+     (origin
+       (inherit (package-source geant4-vis-11-1-1))
```
I wonder: why inherit the source at all when it's overridden with a different source anyway?  (Same for all other instances of `package-source` above and below.)

jakeforster replied:
I use that 1 line to save me these 3:

(package ... (name "geant4")
(origin ... (method git-fetch)
(file-name (git-file-name name version)))

Happy to change it of course.

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                  tcsh
+                  qtbase-5
+                  libxmu
+                  libxt))
```
See above.

jakeforster replied:
Done thanks

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                     ("G4ABLA" ,g4abla-3.3)
+                     ("G4INCL" ,g4incl-1.2)
+                     ("G4ENSDFSTATE" ,g4ensdfstate-2.3)
+                     ("G4TENDL" ,g4tendl-1.4)))))
```
See above.

jakeforster replied:
Done thanks

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+        (base32 "1x7mpi77jylsv8mzsqs0ppchbq147azd0b94i2qq2xhis7m5bn41"))))
+    (build-system gnu-build-system)
+    (native-inputs `(("gcc-toolchain" ,gcc-toolchain)
+                     ("make" ,gnu-make)))
```
These already are default inputs.

jakeforster replied:
I am removing this package from the PR due to license ambiguity.
Thank you for your suggestions nonetheless!

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (source
+     (origin
+       (method url-fetch)
+       (uri "https://geant4.kek.jp/~tanaka/src/dawn_3_91a.tgz")
```
Please use `version` here instead of hardcoding it.

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (native-inputs `(("gcc-toolchain" ,gcc-toolchain)
+                     ("make" ,gnu-make)))
+    (propagated-inputs `(("imagemagick" ,imagemagick)
+                         ("tk" ,tk)))
```
Please drop the labels.

rekado replied:
Do these packages really need to be propagated?

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (propagated-inputs `(("imagemagick" ,imagemagick)
+                         ("tk" ,tk)))
+    (arguments
+     `(#:make-flags (list (string-append "INSTALL_DIR=" %output "/bin"))
```
Please use a G-expression with `#$output`

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+       #:phases (modify-phases %standard-phases
+                  (replace 'configure
+                    (lambda* (#:key outputs #:allow-other-keys)
+                      #t))
```
Please remove this empty phase.  If there's no configure phase, please delete it instead, and add a comment.

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                      #t))
+                  (replace 'check
+                    (lambda* (#:key outputs #:allow-other-keys)
+                      #t))
```
Instead of this, please use `#:tests? #false`.

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                      #t))
+                  (add-after 'build 'create-bin-directory
+                    (lambda* (#:key outputs #:allow-other-keys)
+                      (mkdir-p (string-append %output "/bin")) #t))
```
please use `#$output`, so you can use a simple `(lambda _ ...)`.  Please don't end phases on `#t`.

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+                    (lambda* (#:key outputs #:allow-other-keys)
+                      (substitute* "Makefile"
+                        (("/usr/local/bin")
+                         (string-append %output "/bin"))) #t)))))
```
Same comment as for the previous build phase.

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+which reads 3D geometrical data and visualize them. It is a vectorized
+3D PostScript processor with analytical hidden line/surface removal.
+It aims at precise technical drawing of complex geometries.")
+    (home-page "https://geant4.kek.jp/~tanaka/DAWN/About_DAWN.html")
```
nitpick: I like to have the home-page above the synopsis field.

jakeforster replied:
This package was removed

rekado commented on guix-science/packages/geant4.scm:
```diff
@@ -0,0 +1,699 @@
+    (synopsis "Visualize 3D data generated by Geant4 simulation")
+    (description
+     "Fukui Renderer DAWN (Drawer for Academic WritiNgs) is a renderer,
+which reads 3D geometrical data and visualize them. It is a vectorized
```
Please use two spaces after a period.  (It's an Emacs convention for navigating between sentences.)

jakeforster replied:
This package was removed

rekado reviewed (changes requested):
Thank you for this big patch set!

jakeforster commented:
> Thank you for this big patch set!

Hi rekado

Thank you for your feedback and helpful comments! I've tried to implement your suggestions. I think it's a lot clearer now what is different between geant4 and geant4-vis and what is changing between minor versions of Geant4.

I'm happy to submit this to guix-science-nonfree instead as you suggested. Feel free to close this PR and I can open a new one over there.

I made the following changes in addition to your suggestions:
- only provide a single patch version of Geant4 per minor version
- removed the DAWN package due to license ambiguity
- removed unused module imports

Thanks!
Jake

rekado reviewed (approved):
This looks all pretty good to me.  Thank you for addressing all of my comments!
One minor issue is the indentation throughout.  It looks a little off in some places, but I can fix this when I apply it.

I'll apply this to guix-science-nonfree instead, pending an informed decision (not by me, because I remain uninformed) about the license.

Thanks!

rekado commented:
Added here: https://github.com/guix-science/guix-science-nonfree/commit/26a865d435102fd8f02c9324392752ff51109cd5

(I managed to get the commit summary wrong, but the rest should be fine.)

Thank you for your patience and perseverance!
